### PR TITLE
Paper: Generalized earthquake classification

### PR DESCRIPTION
**Pull request by blasscoc**
Hi Scipy 2016,

This is my submission of the conference proceedings. 

The code that created each of the figures is available on:
https://github.com/blasscoc/FocalMechClassifier
under the "demo" directory. Look for test[2-6].py, run these scripts from the demo directory.

For reproducibility, I've based the results on a dataset that is available with the USGS HASH software, 
http://earthquake.usgs.gov/research/software/HASH/hash.v1.2.tar.gz
Anyone can get this code. 

But I didn't want to start distributing their test data, so you'll have to download HASH and copy the north1.phase and scsn.reverse to the demo directory for the demos to run. 

The version of sklearn I'm using is 0.17, which is the latest available from enpkg. You probably need to go get the Obspy code from their github, I'm using 1.0.1, the version on enpkg is way old.

Thanks
Ben Lasscock

### CODE REVIEW

**Comment by suchow** [comment]
Here's some advice from Proselint (http://github.com/amperser/Proselint), a Python-based linter for prose, with three false alarms removed:

```
18:46: consistency.spacing Inconsistent spacing after period (1 vs. 2 spaces). Found 27 times elsewhere.
61:23: lexical_illusions.misc There's a lexical illusion here: a word is repeated.
382:40: lexical_illusions.misc There's a lexical illusion here: a word is repeated.
66:41: weasel_words.very Substitute 'damn' every time you're inclined to write 'very;' your editor will delete it and the writing will be just as it should be. Found 7 times elsewhere.
```

File: papers/ben_lasscock/00_blasscock.rst [hunk 1]
```diff
@@ -0,0 +1,760 @@
...
+:institution: Geotrace Technologies
+
+-------------------------------------
+Generalized earthquake classification
+-------------------------------------
+
+.. class:: abstract
+
+	   
+   The problem we tackle in the is paper is one of characterizing the
```
Comment (sbenthall) [inline papers/ben_lasscock/00_blasscock.rst]: typo here

File: papers/ben_lasscock/00_blasscock.rst [hunk 2]
```diff
@@ -0,0 +1,760 @@
...
+   source of an earthquake based on identitfying the nodal lines of
```
Comment (sbenthall) [inline papers/ben_lasscock/00_blasscock.rst]: and here

File: papers/ben_lasscock/00_blasscock.rst [hunk 3]
```diff
@@ -0,0 +1,760 @@
...
+   the radiation pattern it produces. These charateristics are the
```
Comment (sbenthall) [inline papers/ben_lasscock/00_blasscock.rst]: and here. you should run a spell cehck

File: papers/ben_lasscock/00_blasscock.rst [hunk 4]
```diff
@@ -0,0 +1,760 @@
...
+   created by three types of earthquakes in terms of a basis of
+   spherical harmonic functions.  The source templates summarized are
+   double couple (D.C.), tensile dislocation (Tensile) and tangential
+   dislocation (Tangential).  The brackets $(\cdot,\cdot)$ define the
+   template direction of the fault normal and direction of slip in
+   rectangular coordinates.  The constant $\alpha = 2 +
+   3\frac{\lambda}{\mu}$, where $\lambda$ and $\mu$ are the first Lame
+   parameter and the shear modulus respectively.} 
+   
+   \end{table}
```
Comment (sbenthall) [inline papers/ben_lasscock/00_blasscock.rst]: As a reader with background in machine learning but not earthquake modeling, this table and its caption are totally opaque to me. Most of the technical terms used have not yet been introduced in the paper. I don't think it's fair to assume they are common knowledge to the general scientific python community.

File: papers/ben_lasscock/00_blasscock.rst [hunk 5]
```diff
@@ -0,0 +1,760 @@
...
+
+   k(\vec{x},\vec{x}_{i}) = (\langle \vec{x}, \vec{x}_{i} \rangle + 1)^{d}\ .
+	 
+Here "d" is the degree of the kernel. The parameter "C" in the above code
+snippet is a slack variable. This allows for a soft thresholding, which allows
+for some misclassification; the default value is usually sufficient.
+   
+.. raw:: latex
+
+   Given a set of data $y_{i}$, which is either 1 or -1, the support
```
Comment (sbenthall) [inline papers/ben_lasscock/00_blasscock.rst]: I don't understand what's meant here, by "Given a set of data $y_{i}$, which is either 1 or -1". This makes it hard for me to understand this whole paragraph, though I think I understand how kernel tricks work generally.

File: papers/ben_lasscock/00_blasscock.rst [hunk 6]
```diff
@@ -0,0 +1,760 @@
...
+to an event from the Northridge dataset. The red line represents zero
+of the classifying function f(x), the green line is the solution for the fault
+(and auxillary) planes determined by HASH. Note that the auxillary
+plane is computed using the aux_plane function provided by the Obspy
+library. The learned nodal line is simply connected, the zero of the
+classifying function f(x) have been determined using matplotlib's contour
+function.
+
+Both the HASH solution and the learned solution have a similar rate of
+misclassification.  However the learned solution is still
```
Comment (sbenthall) [inline papers/ben_lasscock/00_blasscock.rst]: Can you be explicit about the misclassification rate? A comparison of ROC curves, for example, would be useful for evaluating to what extent this new technique is an improvement on HASH.

**Comment by sbenthall** [comment]
Hello,

I decided to try reviewing this submission because I saw there was little activity on this pull request.

I have to admit that much of this paper is over my head as I have no background in seismic waves and sources.

So what I can say is rather superficial:

- The spherical visualizations are worth a thousand words.
- I think you assume a lot of specialized mathematical background. The more you can explain this, or provide references to where the concepts you are using are explained, the better for a general audience.
- There are a number of spelling and grammatical mistakes throughout the paper which are deadly to understanding in a piece that it otherwise so technically dense and precise.

I'm so glad to have read this submission; it's truly humbling to do have done so.

While I could be greatly underestimating the geological expertise available among the other reviewers, at this point I recommend that you narrow the focus of this paper to just that content that surrounds your use of Scientific Python, explain the necessary subset of the earthquake science necessary to understand that very carefully (as if to a more general scientific audience), and submit a paper with your special and novel scientific insights to a conference that's more specific.

**Comment by danlewis85** [comment]
Like @sbenthall I have no background in the area, however a few notes nonetheless.

Generally, this is a good submission, for the most part the meaning is clear and systematically presented. I like the integration of python with the wider context of earthquake classification, and I would suggest leaving it as is as I suspect it will be highly beneficial to your disciplinary peers. As noted previously, the visualisation is high quality and greatly aids the understanding of the text.

I would agree with @sbenthall however in noting that there are a number of lapses of clarity that affect the readers ability to follow the paper, the most severe are:

Abstract - 3rd sentence - incomplete/fragment.

Problem statement - Paragraph 1, 2nd sentence - incomplete/fragment; 7th sentence - incomplete/fragment. Paragraph 4, 2nd sentence -  incomplete/fragment.

Page 5, Column 1, final paragraph - incomplete/fragment.

File: papers/ben_lasscock/00_blasscock.rst [hunk 7]
```diff
@@ -0,0 +1,768 @@
...
+
+.. class:: abstract
+
+	   
+   The problem we tackle in the is paper is one of characterizing the
+   source of an earthquake based on identifying the nodal lines of
+   the radiation pattern it produces. These characteristics are the
+   mode of failure of the rock (shear or tensile), the orientation of
+   the fault plane and direction of slip. We will also extend the
+   derive a correlation coefficient that can be used to compare the
```
Comment (carsonfarmer) [inline papers/ben_lasscock/00_blasscock.rst]: typo here

File: papers/ben_lasscock/00_blasscock.rst [hunk 8]
```diff
@@ -0,0 +1,768 @@
...
+   source mechanisms of different earthquakes.  This information is
+   relevant to both hazard and commercial seismic monitoring.  We will
+   formulate this problem in terms of a simple binary classification
+   on the surface of the sphere, where the classes are the polarity of
+   the first motion at observed various seismic stations, and the
```
Comment (carsonfarmer) [inline papers/ben_lasscock/00_blasscock.rst]: observed at various seismic stations?

**Comment by carsonfarmer** [comment]
One more reviewer with limited domain-specific knowledge. However, some comments on the general submission:
1. The introduction, while setting the stage quite nicely (much stronger than the Abstract), is difficult to follow. This detracts from the submission in general. Examples: (maybe vs may be), and the 2nd to last sentence in the 2nd to last paragraph of the introduction is overly wordy and difficult to follow.
2. I would suggest that a native english speaker do a quick run over this submission before it is made 'final'. There are lots of services that can do this for you.
3. This submission has quite a lot of domain-specific 'jargon'. Consider that Scipy is really for a general scientific audience... some scientific jargon is ok, but submissions should not assume domain-specific knowledge.
4. I agree with @sbenthall, this paper would benefit from separating (to the extent possible) the earthquake science from the scientific Python work. Your reader doesn't need to know too much about earthquake physics to understand SVMs and how they may improve existing classification strategies. A greater focus on methodological improvements and accuracy assessments would likely help a great deal here.
5. Your conclusion is particularly strong. Here you are concise, summarize your findings well, limit domain-specific jargon, and present a good case for your results. I would suggest you use your conclusions as a template for revising the rest of the paper.
6. Great figures!

File: papers/ben_lasscock/00_blasscock.rst [hunk 9]
```diff
@@ -0,0 +1,768 @@
...
+   The correlation score for each event in the Northridge dataset, comparing
+   to event 3146815. (red) The events with maximum and minimum correlation score.
+   (green) The correlation between 3146815 and itself rotated by 90-degrees strike,
+   (cyan) the correlation between 3146815 and the tensile source found in
+   Table 1. :label:`corr`
+
+Finally, we derived a metric of similarity based on a correlation
+score Eq. :ref:`correl`.  To provide an example of how we might use
+this correlation score, we take the event 3146815, which has the
+largest number of data associate with it, and compute the correlation
```
Comment (carsonfarmer) [inline papers/ben_lasscock/00_blasscock.rst]: data points?

File: papers/ben_lasscock/00_blasscock.rst [hunk 10]
```diff
@@ -0,0 +1,768 @@
...
+coefficient with each of the other events in the Northridge
+dataset. According to the reference [HASH]_ the Northridge dataset we
```
Comment (carsonfarmer) [inline papers/ben_lasscock/00_blasscock.rst]: drop: the reference

**Comment by kwinkunks** [comment]
_I apologize in advance for the old school comments here, I just find it easiest. I didn't find any major problems, so most of these are just suggestions — take them or leave them as you wish._

An interesting paper and novel approach to a tricky problem. Well done, and thanks for submitting! It's one of the few instances I can remember where the machine learning problem is so elegantly tied to the physics of the problem. I think you should be careful to emphasize this point when you present, since there will be many (99%) non-seismologists there, and I think it's an interesting philosophical point.

As a general point, I did wonder if you'd tested the approach, and compared to FPFIT, on synthetic data? That seems like a good way to understand its relative performance, since you don't know the actual focal mechanisms for natural events. Maybe someone will ask at the conference.

Consider giving your code a DOI and citing it more robustly in your paper, e.g. with a version number. See https://guides.github.com/activities/citable-code/. This will help others reproduce your work, especially if the code is likely to change in the coming weeks / months.


### Throughout

- It's easier to follow the mathematics and code if you match the typography in the text to that in the display equations and code blocks.
- Prefer proper references to footnotes where it makes sense; see related remarks in 'References'.
- I understand 'scikit-learn' should be written like so, not 'scikits-learn'. 


### Abstract

- Typo in the first line. You can proabably just replace everything up to 'characterizing' with 'We characterize'
- I'd stick an Oxford comma before 'and direction of slip', to make it clear that this is a list of 3 things
- Delete 'extend the' before 'derive'?


### Discussion

- 'we have show' should be 'we have shown'
- 'we analyzing' should be 'we analysed'


### Conclusions

- The wording of the final conclusion is a little hard to parse. Perhaps, 'the correlation coefficient seems to be a reliable measure of the similarity between the focal mechanisms of different events'.


### References

- Good for you for citing the various projects properly. Please consider citing the other software projects you depend on too, especially [ObsPy](https://github.com/obspy/obspy/wiki), which is an academic project, so they'd appreciate it I'm sure. Also Hashpy. Both of these projects have DOIs.
- Some funky indentation.

File: papers/ben_lasscock/00_blasscock.rst [hunk 11]
```diff
@@ -0,0 +1,768 @@
...
+analyzing is expected to contain similar source mechanisms and
```
Comment (carsonfarmer) [inline papers/ben_lasscock/00_blasscock.rst]: are

File: papers/ben_lasscock/00_blasscock.rst [hunk 12]
```diff
@@ -0,0 +1,768 @@
...
+   (top left) The solution for event 3146815, (top right) the solution for
+   events 3158361 and (bottom right) 3153955. Events 3158361 and 3153955
+   represent the maximum and minimum correlation score
+   with event 3146815. :label:`highlo`
+	    
+In :ref:`highlo` we provide a visualization of the events with the
+highest (top right) and lowest (bottom right) correlation score
+comparing with event 3146815 (top left).  The orientation of the nodal
+lines for event 3153955, which has the lowest correlation score,
+indeed is qualitatively very different than solution for
```
Comment (carsonfarmer) [inline papers/ben_lasscock/00_blasscock.rst]: than the solution for

**Comment by sarostru** [comment]
Due to the complication of the proceedings deadline lining up with everyone's travel schedule to attend the conference we've decided to allow authors to spend an additional 2 days, that is until July 12th to complete any major changes to the submission.  This may impact your enjoyment of the conference, so we hope not too many of you need to take the time.  For reviewers, we ask that you read over this final document before the conference is over and post your decision here in the comments on or before July 17th.

Thanks and have a great Scipy!
Your proceedings co-chairs Scott Rostrup (@sarostru) and Sebastian Benthall (@sbenthall)

**Comment by blasscoc** [comment]
Thanks to all of the reviewers, I appreciate you spending time to review my submission.

I've integrated as your suggestions as much as possible. The comments of sbenthall were well taken, and I have certainly written my talk with his comments in mind. However I've kept tone of the manuscript largely unchanged, hopefully improvements in gammer etc will help the understanding.

**Comment by blasscoc** [comment]
kwinkunks originally I did test the code on some simple synthetics. However, minimizing the overall rate of misclassification is important, but isn't really the whole story. The point I try to make is that some misclassification close to the nodal lines is to be expected, the signal to noise is low there, so we shouldn't be too worry about that. Fortunately the NorthRidge dataset had an example that helped demonstrate that.

**Comment by blasscoc** [comment]
Just notice some minor niggles, thought I may as well push them through. I won't make anymore changes.

**Comment by sbenthall** [comment]
Thanks for these edits. It now reads very well. I'd like to recommend this paper for acceptance, with the caveat that I feel like I wouldn't be able to follow the full depth of this paper without studying it carefully for several hours. I would like to hear from other reviewers whether they have any more confidence in their reading of the paper.

**Comment by carsonfarmer** [comment]
Thank you for attending to these edits. I also think, based on my limited knowledge of the subject, this submission is now ready for acceptance.

**Comment by kwinkunks** [comment]
I agree that this submission is good and should be accepted.
